ihmc-java-toolkit/src/main/java/us/ihmc/tools/thread/LoopingThread.java

```diff
@@ -0,0 +1,193 @@
+package us.ihmc.tools.thread;
+
+import us.ihmc.commons.Conversions;
+import us.ihmc.commons.RunnableThatThrows;
+import us.ihmc.commons.exception.ExceptionHandler;
+import us.ihmc.commons.exception.ExceptionTools;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A thread that executes code in a loop.
+ * Optionally, you may set a limit to the loop frequency.
+ * This thread does not finish running until {@link #close()} or {@link #blockingClose()} is called.
+ * Once started, make sure to close this thread.
+ */
+public class LoopingThread extends Thread
```

Comment:
I would recommend not extending thread. It makes a bunch of public erasure and methods that are not part of the API of this class. It should be kept private internally.

Reply:
I prefer that it extends Thread. It makes expectations clearer and you can treat it as standard Java API.

Reply:
Which methods are you thinking of? I looked through the methods that `Thread` offers and I didn't see any that I wouldn't want this class to have, besides the deprecated ones. I'm not seeing the downside to extending `Thread`

Reply:
Well you can't guarantee the thing works because the user can call all those things like yield and interrupt, etc. at any time, and it'll break the assumptions made by the API you added. I would say for the advanced user case, you could accept a user created Thread in the constructor so they could operate on it externally.

Reply:
The main one here is Thread.run(). That's really confusing that that would be available as a method on this class, but you really don't want to be calling that.

Reply:
Yeah, I agree that having `run()` is a bit weird. But you don't really want to call that even with a normal `Thread`. You probably don't want to call `runInLoop()` of this class either. I think it's just a price to pay for enabling the user to extend the class to add custom functionality.

Reply:
I think Java's Thread is pretty badly designed in this respect, but it can't be fixed because it would break too much. I feel pretty strongly that this class should not extend thread. What about providing getThread() to access the underlying thread? Is there a use case I'm missing?

Reply:
I think that'd be fine functionality wise, but I'd find it weird that `LoopingThread` isn't a `Thread`. I'll make that change for now.